File: solutions/cpp/matching-brackets/1/matching_brackets.cpp
#include "matching_brackets.h"

namespace matching_brackets {
    bool check(std::string str) {
        std::stack<char> check_stack;

        for (char c: str) {
            switch (c) {
                case '[': check_stack.push(c); break;
                case '{': check_stack.push(c); break;
                case '(': check_stack.push(c); break;
                case ']':
                    if (check_stack.empty() || check_stack.top() != '[')
                        return false;
                    check_stack.pop(); 
                    break;
                case '}': 
                    if (check_stack.empty() || check_stack.top() != '{')
                        return false;
                    check_stack.pop(); 
                    break;
                case ')': 
                    if (check_stack.empty() || check_stack.top() != '(')
                        return false;
                    check_stack.pop(); 
                    break;
            }
        }
        return check_stack.empty();
    }    
}  // namespace matching_brackets

File: solutions/scala/matching-brackets/1/src/main/scala/MatchingBrackets.scala
import scala.collection.mutable.Stack

object MatchingBrackets {
  def isPaired(brackets: String, idx: Int = 0, stack: Stack[Char] = Stack[Char]()): Boolean = idx match {
    case n if n == brackets.length => stack.isEmpty;
    case n if n < brackets.length => brackets(n) match {
      case c if c == '{' => {
        stack.push('}');
        isPaired(brackets, idx + 1, stack)
      }
      case c if c == '(' => {
        stack.push(')');
        isPaired(brackets, idx + 1, stack)
      }
      case c if c == '[' => {
        stack.push(']');
        isPaired(brackets, idx + 1, stack)
      }
      case c1 if Set('}', ')', ']').contains(c1) => {
        if (stack.isEmpty) false
        else {
          val c2 = stack.pop();
          if (c1 == c2) isPaired(brackets, idx + 1, stack)
          else false
        }
      }
      case _ => isPaired(brackets, idx + 1, stack)
    }
  }
}
